source/extensions/filters/http/oauth2/filter.h

```diff
@@ -48,14 +48,20 @@ class SDSSecretReader : public SecretReader {
   SDSSecretReader(Secret::GenericSecretConfigProviderSharedPtr&& client_secret_provider,
                   Secret::GenericSecretConfigProviderSharedPtr&& token_secret_provider,
                   ThreadLocal::SlotAllocator& tls, Api::Api& api)
-      : client_secret_(std::move(client_secret_provider), tls, api),
-        token_secret_(std::move(token_secret_provider), tls, api) {}
-  const std::string& clientSecret() const override { return client_secret_.secret(); }
-  const std::string& tokenSecret() const override { return token_secret_.secret(); }
+      : client_secret_(
+            THROW_OR_RETURN_VALUE(Secret::ThreadLocalGenericSecretProvider::create(
+                                      std::move(client_secret_provider), tls, api),
+                                  std::unique_ptr<Secret::ThreadLocalGenericSecretProvider>)),
+        token_secret_(
+            THROW_OR_RETURN_VALUE(Secret::ThreadLocalGenericSecretProvider::create(
+                                      std::move(token_secret_provider), tls, api),
+                                  std::unique_ptr<Secret::ThreadLocalGenericSecretProvider>)) {}
+  const std::string& clientSecret() const override { return client_secret_->secret(); }
+  const std::string& tokenSecret() const override { return token_secret_->secret(); }
 
 private:
-  Secret::ThreadLocalGenericSecretProvider client_secret_;
-  Secret::ThreadLocalGenericSecretProvider token_secret_;
+  std::unique_ptr<Secret::ThreadLocalGenericSecretProvider> client_secret_;
```

Comment:
Why did these have to become pointers instead of objects? Would it be unnecessary if `ThreadLocalGenericSecretProvider` had a move-constructor?

Reply:
because the factory create-or-error functions Yan preferred only allow for returning unique pointers :-/

Reply:
Which functions? Couldn't the create function return type be `absl::StatusOr<ThreadLocalGenericSecretProvider>` instead of having a unique_ptr in it?

Reply:
When I tried that with std::move I got a bunch of crashes locally complaining about alignment issues, CI just seg-faulting
https://github.com/envoyproxy/envoy/pull/37292